pulp_container/app/viewsets.py

```diff
@@ -119,7 +124,103 @@ class Meta:
+            repo_info (tuple): Tuple that contains repo type and  Permissions to
+                be checked.
+
+        Returns:
+            List of repositories pk that the current user can view
+
+        """
+        repo_pks = []
+        for key, param in request.query_params.items():
+            if "repository_version" in key:
+                rv = view.get_resource(param, RepositoryVersion)
+                repo = rv.repository.cast()
+                if (
+                    request.user.has_perm(repo_info.push_perm)
+                    and repo.pulp_type == repo_info.push_type
+                ):
+                    repo_pks.append(repo.pk)
+                elif (
+                    request.user.has_perm(repo_info.mirror_perm)
+                    and repo.pulp_type == repo_info.mirror_type
+                ):
+                    repo_pks.append(repo.pk)
+                elif request.user.has_perm(repo_info.push_perm, repo) or request.user.has_perm(
+                    repo_info.mirror_perm, repo
+                ):
+                    repo_pks.append(repo.pk)
+        return repo_pks
+
+    def get_queryset(self):
+        """
+        Gets a QuerySet based on the current request.
+
+        Filters and retuns the only the repo's content user can see.
+
+        Returns:
+            django.db.models.query.QuerySet: The queryset returned contains content the user can see
+            based on the repo permissions
+
+        """
+        qs = super().get_queryset()
+        has_model_push_repo = self.request.user.has_perm(repo_info.push_perm)
+        has_model_repo = self.request.user.has_perm(repo_info.mirror_perm)
+        # this will show also orphaned content
+        if has_model_push_repo and has_model_repo:
+            return qs
+        query_params = self.request.query_params
+        if query_params and "repository_version" in query_params.keys():
+            repo_pks = self._repo_query_params(self.request, self, repo_info)
+            content_qs = qs.model.objects.filter(repositories__in=repo_pks)
+        else:
+            content_type = qs.model.get_pulp_type()
+            repo_qs = models.ContainerRepository.objects.filter(
+                content__pulp_type=content_type, content__in=qs
+            ).only("pk")
+            push_repo_qs = models.ContainerPushRepository.objects.filter(
+                content__pulp_type=content_type, content__in=qs
+            ).only("pk")
+
+            push_repos_can_see = get_objects_for_user(
+                self.request.user, repo_info.push_perm, klass=push_repo_qs
+            ).values_list("pk")
+            mirror_repos_can_see = get_objects_for_user(
+                self.request.user, repo_info.mirror_perm, klass=repo_qs
+            ).values_list("pk")
+            content_qs = qs.model.objects.filter(
+                Q(repositories__in=push_repos_can_see) | Q(repositories__in=mirror_repos_can_see)
+            )
```

Comment:
I agree on the logic here. I wonder if the queries are too heavy because a lot of filtering is happening by field on a different model causing large joins.
Maybe we can get all repos user has access to and then filter content, is it any worse?
E.g.
```
allowed_push_repos = get_objects_for_user(self.request.user, repo_info.push_perm, klass=models.ContainerPushRepository).only('pk')
allowed_mirror_repos = get_objects_for_user(self.request.user, repo_info.push_perm, klass=models.ContainerRepository).only('pk')
content_qs = qs.model.objects.filter(
￼                Q(repositories__in=allowed_push_repos) | Q(repositories__in=allowed_mirror_repos)
```
It's just avoiding searching repos by content. The cost is dealing with more repos in a subsequent query. I think it's likely ok. Any thoughts or do I miss anything?

I'd ask @dralley on the query performance assessment here.

Reply:
I agree.  You can use ```.only("pk")``` just in case if you want.

Reply:
fresh look always helps! i do think it this approach should be better.